### PR TITLE
Major attribute upgrade

### PR DESCRIPTION
**Pull request by Herringway**
Kind of a big PR, but here we go:
- Upgraded many `@system` and `@trusted` functions to `@safe`
- Removed attributes where they can be inferred
- Added attributes to unittests
- Reworked test suite a little to support being mostly `@safe`
- Removed ineffectual `@system` calls (GC disabling, assumeSafeAppend on unsliced arrays)

There is one breaking change that I'm aware of in this PR: dyaml.representer.Representer.addRepresenter now requires a `@safe` function as a parameter.

More can be done, but it would require some larger refactoring that I felt was out of scope here. Even more will be able to be marked `@safe` in a future version of phobos - the Array container is currently not `@safe`, but it looks like it will be soon.

### CODE REVIEW

**Comment by Herringway** [comment]
Odd. All the dub builds on Travis CI seem to have succeeded, but the meson/ninja builds all failed. I have no experience with the latter system, so I don't know what to do about that...

**Comment by wilzbach** [comment]
Me neither - @ximion is the "maintainer" of this. I recommend commenting the ninja comments out in the Travis config, s.t. this doesn't block your PR.

BTW I also just invited you to the YAML team as you seem to be one of the few users of D-YAML and have done a lot of good work recently. Welcome!

File: source/dyaml/node.d [hunk 1]
```diff
@@ -265,7 +265,7 @@ struct Node
         }
         /// Ditto.
         // Overload for types where we can make this nothrow.
-        this(T)(T value, const string tag = null) @trusted pure nothrow
+        this(T)(T value, const string tag = null) @trusted
```
Comment (wilzbach) [inline source/dyaml/node.d]: IIRC once one attribute is specified manually, the compile won't infer the rest.

File: source/dyaml/parser.d [hunk 2]
```diff
@@ -347,7 +347,7 @@ final class Parser
             while(scanner_.checkToken(TokenID.Directive))
             {
                 const token = scanner_.getToken();
-                const value = token.value;
+                string value = token.value.idup;
```
Comment (wilzbach) [inline source/dyaml/parser.d]: Is the duplication really necessary here?

File: source/dyaml/parser.d [hunk 3]
```diff
@@ -432,7 +432,7 @@ final class Parser
             uint tagHandleEnd;
 
             //Get anchor/tag if detected. Return false otherwise.
-            bool get(const TokenID id, const Flag!"first" first, ref string target)
+            bool get(const TokenID id, const Flag!"first" first, ref string target) @trusted
```
Comment (wilzbach) [inline source/dyaml/parser.d]: That's quite a wide `@trusted`  - can't you use a local @trusted lambda?

File: source/dyaml/test/inputoutput.d [hunk 4]
```diff
@@ -25,10 +25,10 @@ alias std.system.endian endian;
 /// Params:  wrong = Get the incorrect BOM for this system.
 ///
 /// Returns: UTF-16 byte order mark.
-wchar bom16(bool wrong = false) pure
+wchar bom16(bool wrong = false) pure @safe
 {
-    wchar little  = *(cast(wchar*)ByteOrderMarks[BOM.UTF16LE]);
-    wchar big     = *(cast(wchar*)ByteOrderMarks[BOM.UTF16BE]);
+    wchar little = '\uFEFF';
+    wchar big = '\uFFFE';
```
Comment (wilzbach) [inline source/dyaml/test/inputoutput.d]: Would be cool if the usage of the global definitions would be kept.

File: source/dyaml/test/reader.d [hunk 5]
```diff
@@ -38,13 +38,13 @@ void runReader(const bool verbose, void[] fileData)
 ///
 /// Params:  verbose       = Print verbose output?
 ///          errorFilename = File name to read from.
-void testStreamError(bool verbose, string errorFilename)
+void testStreamError(bool verbose, string errorFilename) @trusted
```
Comment (wilzbach) [inline source/dyaml/test/reader.d]: Use `readText.representation` instead?

File: source/dyaml/test/representer.d [hunk 6]
```diff
@@ -24,7 +24,7 @@ import dyaml.test.constructor;
 ///          codeFilename = File name to determine test case from.
 ///                         Nothing is read from this file, it only exists
 ///                         to specify that we need a matching unittest.
-void testRepresenterTypes(bool verbose, string codeFilename)
+void testRepresenterTypes(bool verbose, string codeFilename) @trusted
```
Comment (wilzbach) [inline source/dyaml/test/representer.d]: better mark the unittest as `@system` is this isn't safe yet.

**Comment by Herringway** [comment]
looks like the fix was actually pretty simple. marking dyaml.node.Pair.cmp with `@safe` did the trick, but since it's a template, it should have been automatically inferred (and is, but only when dub is used?) so it smells like a dmd frontend bug to me

File: source/dyaml/test/representer.d (hunk 6 above)
Comment (Herringway) [reply source/dyaml/test/representer.d]: The only thing `@system` about this function is a cast(string)ubyte[], and the result of the cast is only used if the tests are already failing. I don't believe there is any chance for corruption or undefined behaviour here.

File: source/dyaml/test/reader.d (hunk 5 above)
Comment (Herringway) [reply source/dyaml/test/reader.d]: No good. Some of the tests use invalid UTF8, and readText reads AND validates.

File: source/dyaml/test/representer.d (hunk 6 above)
Comment (wilzbach) [reply source/dyaml/test/representer.d]: My point was that you have a test that is @safe to ensure that the inference works correctly, but then you set the wrapping tester to `@trusted`.

File: source/dyaml/parser.d (hunk 2 above)
Comment (Herringway) [reply source/dyaml/parser.d]: for now, I believe so. There's a lot of immutable shenanigans that I'd like to eliminate in a later PR, but for now, this didn't seem to be affecting performance in my testing.

File: source/dyaml/node.d (hunk 1 above)
Comment (Herringway) [reply source/dyaml/node.d]: Double checked this just now - it doesn't appear to be the case.

File: source/dyaml/parser.d (hunk 3 above)
Comment (Herringway) [reply source/dyaml/parser.d]: I probably should have .idup'd instead to be consistent...

File: source/dyaml/test/inputoutput.d (hunk 4 above)
Comment (Herringway) [reply source/dyaml/test/inputoutput.d]: I could not come up with a way to convert them to wchar/dchar in a `@safe` way, unfortunately. I note that the existing global definitions exist in phobos as well, so I might clean that up in a following PR.

**Comment by ximion** [comment]
@Herringway @wilzbach Jup, that issue had nothing to do with Meson/ninja, and indeed looks like a (minor) issue in the dmd frontend to me (apparently split builds act differently here than union builds, which is definitely not what should happen). But the fix fortunately was relatively easy :-)

**Pull request merged**